contract_variable_quantity/views/contract.xml

```diff
@@ -0,0 +1,38 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <record id="contract_contract_form_view" model="ir.ui.view">
+        <field name="model">contract.contract</field>
+        <field name="inherit_id" ref="contract.contract_contract_form_view" />
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='code']" position="after">
+                <field name="skip_zero_qty" />
+            </xpath>
+            <xpath
+                expr="//field[@name='contract_line_ids']/tree/field[@name='quantity']"
+                position="before"
+            >
+                <field name="qty_type" />
+                <field name="qty_formula_id" invisible="qty_type != 'variable'" />
+            </xpath>
+            <xpath
+                expr="//field[@name='contract_line_ids']/tree/field[@name='quantity']"
+                position="attributes"
+            >
+                <attribute name="invisible">"qty_type != 'fixed'"</attribute>
+            </xpath>
+            <xpath
+                expr="//field[@name='contract_line_fixed_ids']/tree/field[@name='quantity']"
+                position="before"
+            >
+                <field name="qty_type" />
+                <field name="qty_formula_id" invisible="qty_type != 'variable'" />
```

Comment:
in tree view, if you want to make a column invisible you have to put column_invisible, please check this for the rest of fields in tree view